src/coreclr/vm/codeman.cpp

```diff
@@ -694,22 +688,40 @@ ExecutionManager::ReaderLockHolder::~ReaderLockHolder()
     }
     CONTRACTL_END;
 
-    FastInterlockDecrement(&m_dwReaderCount);
-    DecCantAllocCount();
+    int count;
 
-    EE_LOCK_RELEASED(GetPtrForLockContract());
-}
+    do
+    {
+        count = h->count;
+    }
+    while (count != InterlockedCompareExchangeT((int*)&(h->count), count-1, count));
 
-//---------------------------------------------------------------------------------------
-//
-// Returns whether the reader lock is acquired
+    if (count == 1)// h->count == 0
+    {
+        //we are here in relatively rare case
```

Comment:
If we are in a NoHostCalls situation, we should not delete anything.

Reply:
OK, I can leave this code only in wlh destructor (as writers are about to need allocate or delete something). Now this code is both in rlh and wlh destructors.
//TODO

Reply:
There's no reason to do that. Just disable the code when running in NoHostCalls.